### PR TITLE
enable receive side scaling with better metrics

### PR DESCRIPTION
**Pull request by drunkirishcoder**

### CODE REVIEW

**Comment by zeeshanlakhani** [comment]
After discussion w/ @drunkirishcoder and others, the thought is maybe it's not so useful at moment (the number itself), though the characteristic can be helpful for debugging. Something to think upon.

File: ffi/build.rs [hunk 1]
```diff
@@ -26,11 +26,9 @@ fn main() {
         .header("src/shim.h")
         .generate_comments(true)
         .generate_inline_functions(true)
-        .whitelist_type(r"(rte|cmdline|ether|eth|arp|vlan|vxlan)_.*")
-        .whitelist_function(r"(_rte|rte|cmdline|lcore|ether|eth|arp|is)_.*")
-        .whitelist_var(
-            r"(RTE|CMDLINE|ETHER|ARP|VXLAN|BONDING|LCORE|MEMPOOL|ARP|PKT|EXT_ATTACHED|IND_ATTACHED|lcore|rte|cmdline|per_lcore)_.*",
-        )
+        .whitelist_type(r"(rte|eth|ether)_.*")
+        .whitelist_function(r"(_rte|rte|eth|ether|lcore)_.*")
```
Comment (zeeshanlakhani) [inline ffi/build.rs]: as mentioned in chat, we may not need these lcore/per_lcore whitelists, but maybe we can clean that up later once we're sure.

File: core/src/dpdk/mod.rs [hunk 2]
```diff
@@ -144,6 +149,10 @@ impl fmt::Debug for CoreId {
     }
 }
```
Comment (zeeshanlakhani) [inline core/src/dpdk/mod.rs]: line doesn't show up here, but the only lcore functionality we're using now (other than `master_lcore` in name) is https://github.com/williamofockham/nb2/blob/a498b66abdbc93192e6700646bccfb96ec99b155/core/src/dpdk/mod.rs#L117, and I'm wondering if that's something we need or is there another function like socket_id by idx that can be used?

File: ffi/build.rs (hunk 1 above)
Comment (drunkirishcoder) [reply ffi/build.rs]: yeah once all the calls to them are gone, we can remove them.

File: core/src/dpdk/port.rs [hunk 3]
```diff
@@ -411,7 +462,21 @@ impl<'a> PortBuilder<'a> {
...
     pub fn finish(&mut self, promiscuous: bool, multicast: bool, with_kni: bool) -> Result<Port> {
         let len = self.cores.len() as u16;
-        let conf = ffi::rte_eth_conf::default();
+        let mut conf = ffi::rte_eth_conf::default();
+
+        // turns on receive side scaling if port has multiple cores.
+        if len > 1 {
+            conf.rxmode.mq_mode = ffi::rte_eth_rx_mq_mode::ETH_MQ_RX_RSS;
+            conf.rx_adv_conf.rss_conf.rss_hf =
+                (ffi::ETH_RSS_IP | ffi::ETH_RSS_TCP | ffi::ETH_RSS_UDP | ffi::ETH_RSS_SCTP) as u64
```
Comment (zeeshanlakhani) [inline core/src/dpdk/port.rs]: one additional comment, do we want to drive this listing via function, or constant?

File: core/src/dpdk/mod.rs (hunk 2 above)
Comment (drunkirishcoder) [reply core/src/dpdk/mod.rs]: yeah we should be able to find an equivalent function in `numa.h`. I will take a look and if I find one I can add it to the next pr.

File: core/src/dpdk/port.rs (hunk 3 above)
Comment (drunkirishcoder) [reply core/src/dpdk/port.rs]: which function?

Comment (zeeshanlakhani) [reply core/src/dpdk/port.rs]: Moreso a top-level way to grab this set.

**Pull request merged**